src/common/transformations/src/transformations/common_optimizations/fuse_moe.cpp

```diff
@@ -0,0 +1,381 @@
+                            config.scale_type,
+                            ", up: ",
+                            consts.ups[1]->get_element_type());
+            OPENVINO_ASSERT(consts.downs[1] && consts.downs[1]->get_element_type() == config.scale_type,
+                            "precision of down scale must be same with gate, gate:",
+                            config.scale_type,
+                            ", down: ",
+                            consts.downs[1]->get_element_type());
+        }
+        if (consts.gates[2]) {
+            config.zp_type = consts.gates[2]->get_element_type();
+            OPENVINO_ASSERT(consts.ups[2] && consts.ups[2]->get_element_type() == config.zp_type,
+                            "precision of up zp must be same with gate, gate: ",
+                            config.zp_type,
+                            ", up: ",
+                            consts.ups[2]->get_element_type());
+            OPENVINO_ASSERT(consts.downs[2] && consts.downs[2]->get_element_type() == config.zp_type,
+                            "precision of down zp must be same with gate, gate:",
+                            config.zp_type,
+                            ", down: ",
+                            consts.downs[2]->get_element_type());
+        }
+        return expert_no;
+    };
+
+    return std::make_tuple(result, extract_expert);
+}
+
+}  // namespace
+
+ov::pass::FuseMOEExpert::FuseMOEExpert() {
+    MATCHER_SCOPE(FuseMOE);
+
+    auto expert_mask = makePattern(ov::Rank(3));
+    // shape: [batch * seq_len, hidden_dim]
+    auto final_hidden_states = makePattern<ov::op::internal::MOE>({pattern::any_input(), pattern::any_input()});
+    // shape: [1, batch * seq_len, hidden_dim]
+    auto hidden_states = makePattern(ov::Rank(3));
+    // shape: [self.topk * batch, 1]
+    auto routing_weights = makePattern(ov::Rank(2));
+    auto [result, extract_func] = gen_expert_pattern(final_hidden_states, hidden_states, expert_mask, routing_weights);
+
+    matcher_pass_callback callback =
+        [OV_CAPTURE_CPY_AND_THIS, result = result, extract_func = extract_func](ov::pass::pattern::Matcher& m) {
+            op::internal::MOE::Config config;
+            op::internal::MOE::ConstsPerExpert consts;
+            auto expert_no = extract_func(m, config, consts);
+            if (expert_no < 0)
+                return false;
+
+            const auto& pattern_map = m.get_pattern_value_map();
+            auto last_node = pattern_map.at(result).get_node_shared_ptr();
+
+            // [final_hidden_states, hidden_states, expert_mask, routing_weights]
+            auto prev_moe = pattern_map.at(final_hidden_states).get_node_shared_ptr();
+            auto moe = ov::as_type_ptr<op::internal::MOE>(prev_moe);
+            OPENVINO_ASSERT(config == moe->get_config(), "each expert config must be same");
+            moe->add_consts(static_cast<size_t>(expert_no), consts);
```

Comment:
How long time taken for this? If it is time consuming, there is room to concat thoese weights and then to be copied in parallel at GPU.